locator/tablets.cc

```diff
@@ -1339,6 +1337,52 @@ void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr
     tablet_logger.debug("[assert_rf_rack_valid_keyspace]: Keyspace '{}' has been verified to be RF-rack-valid", ks);
 }
 
+void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr tmptr, const abstract_replication_strategy& ars) {
+    assert_rf_rack_valid_keyspace(ks, tmptr, ars,
+        tmptr->get_topology().get_datacenter_racks(),
+        [&tmptr] (host_id host) {
+            return tmptr->is_normal_token_owner(host);
+        }
+    );
+}
+
+void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr tmptr, const abstract_replication_strategy& ars, rf_rack_topology_operation op) {
+    auto dc_rack_map = tmptr->get_topology().get_datacenter_racks();
+
+    switch (op.tag) {
+        case rf_rack_topology_operation::type::add:
+            // include the new node only if it's added to an existing DC.
+            if (dc_rack_map.contains(op.dc)) {
+                tablet_logger.debug("[assert_rf_rack_valid_keyspace]: Including new node {} in DC '{}' and rack '{}'", op.node_id, op.dc, op.rack);
+                dc_rack_map[op.dc][op.rack].insert(op.node_id);
+            }
+            break;
+        case rf_rack_topology_operation::type::remove:
+            // remove the node from the map, if it exists.
+            tablet_logger.debug("[assert_rf_rack_valid_keyspace]: Excluding removed node {} from DC '{}' and rack '{}'", op.node_id, op.dc, op.rack);
+            if (dc_rack_map.contains(op.dc) && dc_rack_map[op.dc].contains(op.rack)) {
+                dc_rack_map[op.dc][op.rack].erase(op.node_id);
+                if (dc_rack_map[op.dc][op.rack].empty()) {
+                    dc_rack_map[op.dc].erase(op.rack);
+                    if (dc_rack_map[op.dc].empty()) {
+                        dc_rack_map.erase(op.dc);
+                    }
+                }
+            }
+            break;
+    }
+
+    assert_rf_rack_valid_keyspace(ks, tmptr, ars,
+        dc_rack_map,
+        [&tmptr, &op] (host_id host) {
+            if (op.tag == rf_rack_topology_operation::type::add && host == op.node_id) {
+                return true;
+            }
+            return tmptr->is_normal_token_owner(host);
+        }
```

Comment:
If I understand correctly, this lambda will return `true` for the new node that is joining - even if the node is a zero token node. So, this looks incorrect to me. However, I can see that you added tests in later commits which verify this. Why does it work?